smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGenerator.kt

```diff
@@ -208,16 +212,17 @@ abstract class HttpProtocolClientGenerator(
         val hasOutputStream = outputShape.map { it.hasStreamingMember(ctx.model) }.orElse(false)
         val inputVariableName = if (inputShape.isPresent) "input" else KotlinTypes.Unit.fullName
 
-        if (hasOutputStream) {
-            writer
-                .addImport(RuntimeTypes.Http.Operation.execute)
-                .write("return op.#T(client, #L, block)", RuntimeTypes.Http.Operation.execute, inputVariableName)
-        } else {
-            writer.addImport(RuntimeTypes.Http.Operation.roundTrip)
-            if (outputShape.isPresent) {
-                writer.write("return op.#T(client, #L)", RuntimeTypes.Http.Operation.roundTrip, inputVariableName)
-            } else {
-                writer.write("op.#T(client, #L)", RuntimeTypes.Http.Operation.roundTrip, inputVariableName)
+        writer.withBlock("return rootTraceSpan.child(op.sdkRequestId).#T { traceSpan ->", "}", RuntimeTypes.IO.use) {
```

Comment:
I wonder if the child span should include the operation name and the request id (e.g. such that the resulting log is like `DynamoDb/PutItem/<request-id>` 🤔

Reply:
Spans are context boundaries which have a meaningful beginning and end. For the entire client, the beginning and end are clear: the lifetime of the client itself, which begins at instantiation and ends when `close` is called. For requests, the beginning and end align with the lifetime of the request and its various middleware. These boundaries are important (especially the end) because probe implementations may collect events _until_ the end of a span before calculating any additional metadata or emitting them downstream.

For a child span that's defined by operation name, what are the beginning and end conditions? It seems like they could only be coterminous with an individual request or (more likely) the client itself. Neither case seems to add incremental benefit for event batching.

Reply:
Yes I suppose `sdkRequestId` and `OperationName` have the same lifetime. I was thinking of this mostly through a debugging lens. Being able to only emit events for a specific operation would be easier with the operation name whereas request ID is dynamic per invocation obviously. It would be harder to filter spans of interest (you either get all of a specific client or all of a specific request ID, nothing in-between).

Reply:
I suppose it's possible the span IDs for requests could combine the operation name and a unique request ID:

Example spans:
* `s3`
  * `listObjects:89e1cbef-00f5-4005-9863-9d18aff6ac52`
    * `attempt 0`
      * ...

That said, I worry we're conflating the logical scope/lifestime of spans with additional information that may be useful for debugging, binning, or filtering. I wonder if we're missing a design concept of additional metadata (or what [CloudWatch calls "dimensions"](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Dimension)) which could be better suited to, for example, probes filtering events to specific operation names.